### DISCUSSION TITLE
Should we keep all Jupyter notebooks?

### DISCUSSION THREAD

**Discussion by jcohenadad** [neuropoly/axondeepseg#148] [Bug reports, suggestions for code refactoring]
Currently, there are 18 notebooks:
- some with very cryptic names
- most of them without documentation (and some doc is in French)
- with some redundancy between notebooks
- most of them contain only few lines of code that don't deserve to be in a notebook for someone who knows a minimum of Python programming

Those notebooks are becoming a dump of "nice functionalities to have" without proper harmonization/guideline for creating them. We should control it otherwise we will end up with 300 notebooks in 10 years.

I see two approaches (for our users):

1.  We abandon notebooks and we implement useful functionalities as a wrapper that would be called from the Terminal, without the need to invoke (and install) Jupyter. Those functionalities could include those done by: generate_axons_from_myelin.ipynb, get_masks, etc.  With that approach, we could build pipelines in bash. 

2. We keep Jupyter notebooks, not only as a "tutorials", but rather as "official tools part of ADS that are necessary for doing XYZ tasks". If we go that route, I suggest we (i) include Jupyter as a requirement and (ii) we do some cleaning on the existing notebooks, with a clear vision about what notebooks are needed, for what purpose, etc.

**Comment by mathieuboudreau** [comment]
I'm leaning on option 1. Jupyter notebooks being "official tools part of ADS that are necessary for doing XYZ tasks" is not appealing to me, as I've had kernels crash on me before, which are a major nuisance. Jupyter notebooks also can't be regularly tested using any testing suites, leaving it up to the individuals to keep it up to date, otherwise you'll encounter issues like #147 .

Notebooks are nice though for new users though. If we do Option 1, I think we'll need to document (with visuals) How-to's somewhere else.

**Comment by alzaia** [comment]
There are definitely a lot of notebooks out there and most of them are 'draft' versions from past students. 

In case you go with approach 2 above, here are the notebooks that are clean enough and well documented:

* The getting started notebook (Binder): https://github.com/neuropoly/axondeepseg/blob/master/notebooks/getting_started.ipynb
* Creating dataset of patches for training: https://github.com/neuropoly/axondeepseg/blob/master/notebooks/guide_dataset_building.ipynb
* Training a model: https://github.com/neuropoly/axondeepseg/blob/master/notebooks/training_guideline.ipynb
* Computing segmentation metrics: https://github.com/neuropoly/axondeepseg/blob/master/notebooks/performance_metrics.ipynb
* Extracting morphometrics: https://github.com/neuropoly/axondeepseg/blob/master/notebooks/morphometrics_extraction.ipynb
* Effect of noise addition: https://github.com/neuropoly/axondeepseg/blob/master/notebooks/noise_simulation.ipynb

**Comment by perone** [comment]
Another alternative is to keep only a minimum of notebooks showing the API usage (like the ones that Aldo posted in the message above), but move everything else to command-line tools, so there is a non-programmatic way to do every step of the pipeline:

1. Build the dataset
2. Training/fine-tuning
3. Prediction (generate prediction masks)
4. Morphometrics extraction from the prediction masks

**Comment by mathieuboudreau** [comment]
@perone The "minimum of notebooks" you suggest basically cover all of the functionality available on AxonDeepSeg API in its current state, so if we do that, I agree with Julien that we should give up the CLI and force the users to use the notebooks for processing. Otherwise, notebooks are neglected during development and get outdated when changes are made, since they cannot be tested with continuous integration.

Currently, the only CLI command we have anyways is for step 3, so it's not like it's a huge loss, I'm just worried notebooks being outdated. But if we're "forced" to use notebooks, and all users and devs use them, then hopefully we will inherently be the continuously testing them ourselves, instead of letting them get stale.

**Comment by jcohenadad** [comment]
there might be ways to include notebooks in our CI:

- https://github.com/ruaridhw/jupyter-ci
- https://www.cityscience.com/blog/ci-for-jupyter.html

**Comment by perone** [comment]
What about if we just keep a single notebook with an example since the data generation up to the morphometrics extraction and execute it with `nbconvert` during the continuous integration ?

I think that if we have a unit testing for the API, that uses the same API used by notebooks, there is no way that the notebooks would fail if the unit testing isn't failing right ?

**Comment by mathieuboudreau** [comment]
>I think that if we have a unit testing for the API, that uses the same API used by notebooks, there is no way that the notebooks would fail if the unit testing isn't failing right ?

That holds unless there are major refactoring, e.g. our previous Python 2 to 3 refactoring and the upcoming planned refactoring. For these cases, tests need to be changed too, most likely.

Also, although I try my best, I don't think we've fully embraced the [semantic versioning](https://semver.org) approach in this project yet, where we we should not allow any API (e.g. function args) changes for \*.x\.* and \*.\*.x versions. That should be embraced too, if we do move forward with Jupyter Notebooks.

**Comment by mathieuboudreau** [comment]
>What about if we just keep a single notebook with an example since the data generation up to the morphometrics extraction and execute it with nbconvert during the continuous integration ?

That's fine, we could even have 3 notebooks (1, 3, 4). 2 is somewhat tricky, as for it to work in any type of CI, we have to change the training config to be something that processes quite quickly, and probably wouldn't be a recommended way, so the content within the notebook would not be as useful for users wanting to do a real training set. (Unless the real config & datasets are commented out, and the notebook says to switch them, but that's not super clean either).

**Comment by antmoev** [comment]
Although I love Jupyter notebooks, I'll keep only one for showing off/education purposes. More notebooks means more maintenance and dependencies over other packages.
